analyzer/protocol/stun/stun.spicy

```diff
@@ -0,0 +1,215 @@
+	EVEN_PORT = 0x0018,
+	REQUESTED_TRANSPORT = 0x0019,
+	DONT_FRAGMENT = 0x001A,
+	ACCESS_TOKEN = 0x001B,
+	MESSAGE_INTEGRITY_SHA256 = 0x001C,
+	PASSWORD_ALGORITHM = 0x001D,
+	USERHASH = 0x001E,
+	XOR_MAPPED_ADDRESS = 0x0020,
+	RESERVATION_TOKEN = 0x0022,
+	PRIORITY = 0x0024,
+	USE_CANDIDATE = 0x0025,
+	PADDING = 0x0026,
+	RESPONSE_PORT = 0x0027,
+	CONNECTION_ID = 0x002A,
+	ADDITIONAL_ADDRESS_FAMILY = 0x8000,
+	ADDRESS_ERROR_CODE = 0x8001,
+	PASSWORD_ALGORITHMS = 0x8002,
+	ALTERNATE_DOMAIN = 0x8003,
+	ICMP = 0x8004,
+	SOFTWARE = 0x8022,
+	ALTERNATE_SERVER = 0x8023,
+	TRANSACTION_TRANSMIT_COUNTER = 0x8025,
+	CACHE_TIMEOUT = 0x8027,
+	FINGERPRINT = 0x8028,
+	ICE_CONTROLLED = 0x8029,
+	ICE_CONTROLLING = 0x802A,
+	RESPONSE_ORIGIN = 0x802B,
+	OTHER_ADDRESS = 0x802C,
+	ECN_CHECK = 0x802D,
+	THIRD_PARTY_AUTHORIZATION = 0x802E,
+	MOBILITY_TICKET = 0x8030,
+	CISCO_STUN_FLOWDATA = 0xC000,
+	ENF_FLOW_DESCRIPTION = 0xC001,
+	ENF_NETWORK_STATUS = 0xC002,
+	GOOG_NETWORK_INFO = 0xC057,
+	GOOG_LAST_ICE_CHECK_RECEIVED = 0xC058,
+	GOOG_MISC_INFO = 0xC059,
+	GOOG_OBSOLETE_1 = 0xC05A,
+	GOOG_CONNECTION_ID = 0xC05B,
+	GOOG_DELTA = 0xC05C,
+	GOOG_DELTA_ACK = 0xC05D,
+	GOOG_MESSAGE_INTEGRITY_32 = 0xC060,
+	};
+
+#       RFC5389:
+#
+#		Magic cookie: 0x2112A442
+#
+#       0                   1                   2                   3
+#       0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+#      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+#      |0 0|     STUN Message Type     |         Message Length        |
+#      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+#      |                         Magic Cookie                          |
+#      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+#      |                                                               |
+#      |                     Transaction ID (96 bits)                  |
+#      |                                                               |
+#      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+#      |                        [Attributes,                           |
+#      |                            ...]                               |
+#      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+#
+# 		msg_type:
+#                       0                 1
+#                       2  3  4 5 6 7 8 9 0 1 2 3 4 5
+#                      +--+--+-+-+-+-+-+-+-+-+-+-+-+-+
+#                      |M |M |M|M|M|C|M|M|M|C|M|M|M|M|
+#                      |11|10|9|8|7|1|6|5|4|0|3|2|1|0|
+#                      +--+--+-+-+-+-+-+-+-+-+-+-+-+-+
+
+public type STUNPacket = unit {
+	var M: uint16;
+	var C: uint16;
+
+	msgtype: bitfield(16) {
+	  m0: 0..3;
+	  c0: 4;
+	  m1: 5..7;
+	  c1: 8;
+	  m2: 9..13;
+	  zeros: 14..15 &requires=($$ == 0);
+	};
+
+	on msgtype {
+		self.M = self.msgtype.m0 + 16*self.msgtype.m1 + 128*self.msgtype.m2;
+		self.C = self.msgtype.c0 + 2*self.msgtype.c1;
+	}
```

Comment:
We don't need `msgtype` after filling in `M` and `C`, let's remove that additional state,

```suggestion
	: bitfield(16) {
	  m0: 0..3;
	  c0: 4;
	  m1: 5..7;
	  c1: 8;
	  m2: 9..13;
	  zeros: 14..15 &requires=($$ == 0);
	} {
		self.M = self.msgtype.m0 + 16*self.msgtype.m1 + 128*self.msgtype.m2;
		self.C = self.msgtype.c0 + 2*self.msgtype.c1;
	}
```

Reply:
I don't know what you mean.

Reply:
This actually broke the code.  I had to add msgtype: back in before the colon.